packages/forms/src/UIForm/fields/NestedListView/NestedListView.utils.js

```diff
@@ -15,25 +15,34 @@ function hasChildren(item) {
  * @returns {Array}
  */
 export function getDisplayedItems(items, value, searchCriteria) {
-	let textFilter;
-	if (searchCriteria) {
-		textFilter = ({ label }) => label.toLowerCase().includes(searchCriteria.toLowerCase());
-	}
+	const textFilter = searchCriteria ? searchCriteria.toLowerCase() : '';
 
-	return items
-		.map(item => {
+	const checkedItems = items
+		.reduce((acc, item) => {
 			const newChildren = item.children.map(child => ({
 				...child,
 				checked: (value[item.key] || []).includes(child.value),
 			}));
-
-			return {
-				...item,
-				checked: newChildren.some(child => child.checked),
-				children: textFilter ? newChildren.filter(textFilter) : newChildren,
-			};
-		})
+			return [
+				...acc,
+				{ ...item, checked: newChildren.some(child => child.checked), children: newChildren },
+			];
+		}, [])
 		.filter(hasChildren);
+
+	return checkedItems.reduce((filtered, item) => {
+		if (item.label.toLowerCase().includes(textFilter)) {
+			return [...filtered, item];
+		}
+
+		const filteredChildren = item.children.filter(child =>
+			child.label.toLowerCase().includes(textFilter),
+		);
+
+		return filteredChildren.length > 0
+			? [...filtered, { ...item, children: filteredChildren }]
+			: filtered;
+	}, []);
```

Comment:
why do we need to reduce twice ? Look quite heavy to reduce a datastructure to then reduce it again.
I am looking forward to the test to look deeper into your algorithm here.
AFAIK we can have lots  of items here, so this is important to stay efficient in this kind of filter